File: python/Maximum Subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        answer = nums[0]
        dp = [0] * len(nums)
        dp[0] = answer
        
        for i in range(1, len(nums)):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
            answer = max(answer, dp[i])
            
        return answer

File: cpp/Maximum Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }
        vector<int> dp(nums.size(), 0);
        dp[0] = nums[0];
        int answer = dp[0];
        for (int i = 1; i < nums.size(); i++) {
            dp[i] = max(nums[i], dp[i - 1] + nums[i]);
            answer = max(answer, dp[i]);
        }
        return answer;
    }
};
